server.go

```diff
@@ -1016,211 +1068,116 @@ type openChanReq struct {
-			s.addPeer(p)
-
-		// Finished peers.
-		case p := <-s.donePeers:
-			s.removePeer(p)
-
-		case bMsg := <-s.broadcastRequests:
-			ignore := bMsg.ignore
-
-			srvrLog.Debugf("Broadcasting %v messages", len(bMsg.msgs))
-
-			// Launch a new goroutine to handle the broadcast
-			// request, this allows us process this request
-			// asynchronously without blocking subsequent broadcast
-			// requests.
-			go func() {
-				s.peersMtx.RLock()
-				for _, sPeer := range s.peersByPub {
-					if ignore != nil &&
-						sPeer.addr.IdentityKey.IsEqual(ignore) {
-
-						srvrLog.Debugf("Skipping %v in broadcast",
-							ignore.SerializeCompressed())
-
-						continue
-					}
-
-					go func(p *peer) {
-						for _, msg := range bMsg.msgs {
-							p.queueMsg(msg, nil)
-						}
-					}(sPeer)
-				}
-				s.peersMtx.RUnlock()
-
-				bMsg.errChan <- nil
-			}()
-		case sMsg := <-s.sendRequests:
-			// TODO(roasbeef): use [33]byte everywhere instead
-			//  * eliminate usage of mutexes, funnel all peer
-			//    mutation to this goroutine
-			target := sMsg.target.SerializeCompressed()
-
-			srvrLog.Debugf("Attempting to send msgs %v to: %x",
-				len(sMsg.msgs), target)
-
-			// Launch a new goroutine to handle this send request,
-			// this allows us process this request asynchronously
-			// without blocking future send requests.
-			go func() {
-				s.peersMtx.RLock()
-				targetPeer, ok := s.peersByPub[string(target)]
-				if !ok {
-					s.peersMtx.RUnlock()
-					srvrLog.Errorf("unable to send message to %x, "+
-						"peer not found", target)
-					sMsg.errChan <- errors.New("peer not found")
-					return
-				}
-				s.peersMtx.RUnlock()
-
-				sMsg.errChan <- nil
-
-				for _, msg := range sMsg.msgs {
-					targetPeer.queueMsg(msg, nil)
-				}
-			}()
-		case query := <-s.queries:
-			switch msg := query.(type) {
-			case *disconnectPeerMsg:
-				s.handleDisconnectPeer(msg)
-			case *connectPeerMsg:
-				s.handleConnectPeer(msg)
-			case *openChanReq:
-				s.handleOpenChanReq(msg)
-			}
-		case <-s.quit:
-			break out
-		}
-	}
-
-	s.connMgr.Stop()
-
-	s.wg.Done()
-}
-
-// handleConnectPeer attempts to establish a connection to the address enclosed
-// within the passed connectPeerMsg. This function is *async*, a goroutine will
-// be spawned in order to finish the request, and respond to the caller.
-func (s *server) handleConnectPeer(msg *connectPeerMsg) {
-	addr := msg.addr
+// ConnectToPeer requests that the server connect to a Lightning Network peer
+// at the specified address. This function will *block* until either a
+// connection is established, or the initial handshake process fails.
+// NOTE: This function is safe for concurrent access.
+func (s *server) ConnectToPeer(addr *lnwire.NetAddress, perm bool) error {
 
-	targetPub := string(msg.addr.IdentityKey.SerializeCompressed())
+	targetPub := string(addr.IdentityKey.SerializeCompressed())
 
 	// Ensure we're not already connected to this
 	// peer.
-	s.peersMtx.RLock()
+	s.mu.Lock()
```

Comment:
In this case _I think_ you cannot use the `defer Unlock()` pattern because you must unlock before calling `s.OutboundPeerConnected(nil, conn)`? Would be nice to have a small comment explaining this 😄